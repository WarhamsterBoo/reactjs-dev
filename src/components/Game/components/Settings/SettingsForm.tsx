import React, { useCallback } from "react";
import { InputNumber } from ".";
import { InputsContainer } from "./SettingsForm.styled";

export interface SettingsFormProps {
  onSubmit: () => void;
}

export const SettingsForm: React.FC<SettingsFormProps> = ({ onSubmit }) => {
  const onHandleSubmit = useCallback(
    (ev: React.FormEvent) => {
      ev.preventDefault();
      onSubmit();
    },
    [onSubmit]
  );

  return (
    <InputsContainer>
      <form onSubmit={onHandleSubmit}>
        <fieldset>
          <legend>Game settings</legend>
          <label>
            X dimension:
            <InputNumber />
          </label>
          <label>
            Y dimension:
            <InputNumber />
          </label>
          <label>
            Filling Percentage:
            <InputNumber />
          </label>
          <button>Apply</button>
        </fieldset>
      </form>
    </InputsContainer>
  );
};
