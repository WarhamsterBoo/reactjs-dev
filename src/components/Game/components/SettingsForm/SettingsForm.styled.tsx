import styled from "@emotion/styled";

export const Form = styled.form`
  padding: 10px 10px 10px 10px;

  border: solid 1px black;
  background-color: #011627;

  display: inline-flex;
  flex-direction: row;
  justify-content: space-around;
  align-items: center;
`;

export const Label = styled.label`
  margin-right: 10px;
  margin-left: 10px;

  color: #addb67;
  font-family: Inconsolata, monospace;
`;

export const Legend = styled.legend`
  color: #addb67;
  font-family: Inconsolata, monospace;
`;

export const FieldSet = styled.fieldset`
  border-color: #addb67;
`;
