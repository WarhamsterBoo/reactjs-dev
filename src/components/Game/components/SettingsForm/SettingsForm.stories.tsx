import { action } from "@storybook/addon-actions";
import React from "react";
import { SettingsForm } from "./SettingsForm";

export default { title: "Settings Form Component" };

export const SettingsFormStory: React.FC<{}> = () => {
  return <SettingsForm onSettingsSubmit={action("onSubmit")} />;
};
