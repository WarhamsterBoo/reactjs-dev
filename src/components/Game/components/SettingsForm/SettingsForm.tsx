import React, { FormEvent, useCallback, useState } from "react";
import { Button, InputNumber, Label } from "shared";
import { FieldSet, Form, Legend } from "./SettingsForm.styled";
import { GameSettings } from "commonTypes/game";

export interface SettingsFormProps {
  gameSettings: GameSettings;
  onSettingsSubmit: (settings: GameSettings) => void;
}

export const SettingsForm: React.FC<SettingsFormProps> = ({
  gameSettings,
  onSettingsSubmit,
}) => {
  const [settings, setSettings] = useState<GameSettings>(gameSettings);

  const onHandleInputChange = useCallback((ev: FormEvent<HTMLInputElement>) => {
    const { name, value } = ev.target as HTMLInputElement;

    setSettings((prevSettings) => ({
      ...prevSettings,
      [name]: parseInt(value),
    }));
  }, []);

  const onHandleSubmit = useCallback(
    (ev: FormEvent) => {
      ev.preventDefault();
      onSettingsSubmit({ ...settings });
    },
    [onSettingsSubmit, settings]
  );

  return (
    <Form onSubmit={onHandleSubmit}>
      <FieldSet>
        <Legend>Game settings</Legend>
        <Label>
          X dimension:
          <InputNumber
            value={settings.xDimension}
            onChange={onHandleInputChange}
            name="xDimension"
          />
        </Label>
        <Label>
          Y dimension:
          <InputNumber
            value={settings.yDimension}
            onChange={onHandleInputChange}
            name="yDimension"
          />
        </Label>
        <Label>
          Filling Percentage:
          <InputNumber
            min="0"
            value={settings.fillingPercentage}
            onChange={onHandleInputChange}
            name="fillingPercentage"
          />
        </Label>
        <Button>Apply</Button>
      </FieldSet>
    </Form>
  );
};
