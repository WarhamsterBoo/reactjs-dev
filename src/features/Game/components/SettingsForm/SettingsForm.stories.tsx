import { GameSettings, GameStatus } from "@/features/Game/gameStore";
import { action } from "@storybook/addon-actions";
import React, { useState } from "react";
import { SettingsForm } from "./SettingsForm";

export default { title: "Settings Form Component" };

export const SettingsFormStory: React.FC<{}> = () => {
  const [settings, setSettings] = useState<GameSettings>({
    xDimension: 10,
    yDimension: 10,
    fillingPercentage: 10,
    speed: 10,
    status: GameStatus.Stopped,
  });

  return (
    <SettingsForm
      gameSettings={settings}
      applySettings={action("onSubmit")}
      onSettingsChange={(newSettings) => setSettings(newSettings)}
    />
  );
};
