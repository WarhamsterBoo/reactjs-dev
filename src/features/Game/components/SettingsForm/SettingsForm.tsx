import { Button, InputNumber, Label } from "@/components";
import React, { FormEvent, useCallback, useState } from "react";
import { FieldSet, Form, Legend } from "./SettingsForm.styled";

interface Settings {
  xDimension: number;
  yDimension: number;
  fillingPercentage: number;
}

export interface SettingsFormProps {
  gameSettings: Settings;
  onSettingsSubmit: (settings: Settings) => void;
  applySettings?: () => void;
}

export const SettingsForm: React.FC<SettingsFormProps> = ({
  gameSettings,
  onSettingsSubmit,
  applySettings
}) => {
  const [settings, setSettings] = useState<Settings>({
    ...gameSettings,
    fillingPercentage: Math.round(gameSettings.fillingPercentage * 100),
  });

  const onHandleInputChange = useCallback((ev: FormEvent<HTMLInputElement>) => {
    const { name, value } = ev.target as HTMLInputElement;

    setSettings((prevSettings) => ({
      ...prevSettings,
      [name]: parseInt(value),
    }));
  }, []);

  const onHandleSubmit = useCallback(
    (ev: FormEvent) => {
      ev.preventDefault();
      onSettingsSubmit({
        ...settings,
        fillingPercentage: settings.fillingPercentage / 100,
      });
      if (applySettings) applySettings();
    },
    [onSettingsSubmit, settings]
  );

  return (
    <Form onSubmit={onHandleSubmit}>
      <FieldSet>
        <Legend>Game settings</Legend>
        <Label>
          X dimension:
          <InputNumber
            value={settings.xDimension}
            onChange={onHandleInputChange}
            name="xDimension"
          />
        </Label>
        <Label>
          Y dimension:
          <InputNumber
            value={settings.yDimension}
            onChange={onHandleInputChange}
            name="yDimension"
          />
        </Label>
        <Label>
          Filling Percentage:
          <InputNumber
            min="0"
            value={settings.fillingPercentage}
            onChange={onHandleInputChange}
            name="fillingPercentage"
          />
        </Label>
        <Button>Apply</Button>
      </FieldSet>
    </Form>
  );
};
