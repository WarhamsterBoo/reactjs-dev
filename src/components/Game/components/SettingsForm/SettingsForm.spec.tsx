import React from "react";
import renderer from "react-test-renderer";
import { SettingsForm } from "./SettingsForm";
import { mount } from "enzyme";

describe("SettingsForm", () => {
  it("should render", () => {
    const sut = <SettingsForm onSubmit={() => {}} />;

    expect(renderer.create(sut).toJSON()).toMatchSnapshot();
  });

  it("should change values in inputs", () => {
    const fakeOnSubmit = jest.fn();
    const sut = mount(<SettingsForm onSubmit={fakeOnSubmit} />);

    sut.find(`input[name="xDimension"]`).simulate("change", {
      target: { value: "10", name: "xDimension" },
    });
    sut.find(`input[name="yDimension"]`).simulate("change", {
      target: { value: "10", name: "yDimension" },
    });
    sut.find(`input[name="fillingPercentage"]`).simulate("change", {
      target: { value: "10", name: "fillingPercentage" },
    });

    expect(sut.find(`input[name="xDimension"]`).prop("value")).toBe(10);
    expect(sut.find(`input[name="yDimension"]`).prop("value")).toBe(10);
    expect(sut.find(`input[name="fillingPercentage"]`).prop("value")).toBe(10);
  });
});
