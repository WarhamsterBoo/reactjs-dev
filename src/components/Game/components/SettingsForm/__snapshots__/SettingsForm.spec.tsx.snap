// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsForm should render 1`] = `
<Styled(form)
  onSubmit={[Function]}
>
  <Styled(fieldset)>
    <Styled(legend)>
      Game settings
    </Styled(legend)>
    <Styled(label)>
      X dimension:
      <InputNumber
        name="xDimension"
        onChange={[Function]}
        value={10}
      />
    </Styled(label)>
    <Styled(label)>
      Y dimension:
      <InputNumber
        name="yDimension"
        onChange={[Function]}
        value={10}
      />
    </Styled(label)>
    <Styled(label)>
      Filling Percentage:
      <InputNumber
        min="0"
        name="fillingPercentage"
        onChange={[Function]}
        value={0}
      />
    </Styled(label)>
    <Styled(button)>
      Apply
    </Styled(button)>
  </Styled(fieldset)>
</Styled(form)>
`;
